main.tf

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+/******************************************
+	Scheduled Function Definition
+ *****************************************/
+
+resource "google_cloud_scheduler_job" "job" {
+  provider = "google-beta"
+  name     = "${var.job_name}"
+  description = "${var.job_description}"
+  schedule = "${var.job_schedule}"
+
+  pubsub_target = {
+    topic_name = "projects/${var.project_id}/topics/${module.pubsub_topic.topic}"
+    data = "${var.message_data}"
+  }
+}
+
+/******************************************
+	PubSub Topic Definition
+ *****************************************/
+
+module "pubsub_topic" {
+  source  = "github.com/terraform-google-modules/terraform-google-pubsub"
+  topic   = "${var.topic_name}"
+  project_id = "${var.project_id}"
+}
+
+/******************************************
+	Cloud Function Resource Definitions
+ *****************************************/
+
+resource "google_cloudfunctions_function" "main" {
+  provider = "google-beta"
+  name                  = "${var.name}"
+  source_archive_bucket = "${google_storage_bucket.main.name}"
+  source_archive_object = "${google_storage_bucket_object.main.name}"
+  description           = "${var.function_description}"
+  available_memory_mb   = "${var.function_available_memory_mb}"
+  timeout               = "${var.function_timeout_s}"
+  entry_point           = "${var.function_entry_point}"
+
+  event_trigger {
+    event_type = "google.pubsub.topic.publish"
+    resource   = "${module.pubsub_topic.topic}"
+
+    failure_policy {
+      retry = "${var.function_event_trigger_failure_policy_retry}"
+    }
+  }
+
+  labels                = "${var.function_labels}"
+  runtime               = "${var.function_runtime}"
+  environment_variables = "${var.function_environment_variables}"
+  project               = "${var.project_id}"
+  region                = "${var.region}"
+}
+
+data "archive_file" "main" {
+  type        = "zip"
+  output_path = "${pathexpand("${var.function_source_directory}.zip")}"
+  source_dir  = "${pathexpand("${var.function_source_directory}")}"
+}
+
+resource "random_string" "random_suffix" {
+  length  = 4
+  upper   = "false"
+  special = "false"
+}
+
+resource "google_storage_bucket" "main" {
+  provider = "google-beta"
```

Comment:
It looks like the beta provider is needed across the board on resources created here. Is there also a version constraint (e.g. `~> 2.0.0`). I know specifying the provider within a module is generally discouraged (and you've left that out 👍 ) but if there are provider version specific resources here, I think that's the right use case to include a provider block.

Also, since this pattern relies on the calling module declaring the `google-beta` provider, does it also rely on that provider **not** being aliased since `google-beta` is hardcoded everywhere?

Reply:
I've updated this. The way it should work, is that the user of this module needs to define the google-beta provider, and pass it in to the module. I've updated the example to show this. That also allows me to remove the provider specific pieces, and it just inherits the provider passed in. Other work would have to be done if they want to mix and match beta/non-beta providers.